ARROW-15658: [C++] Parquet pushdown filtering fails if the filter expression uses numeric field references

vibhatha opened the pull request:
This PR contains an initial step towards a fix for ARROW-15658

- [x] Added a fix to enable numeric field references
- [x] Adopted a test case from the [JIRA description](https://issues.apache.org/jira/browse/ARROW-15658) by @westonpace

westonpace commented:
I think you accidentally pushed some submodule changes.

lidavidm commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -160,18 +160,30 @@ Status ResolveOneFieldRef(
+
+  if (const std::string* name = field_ref.name()) {
+    RETURN_NOT_OK(resolve_field_ref(name));
     // "Virtual" column: field is not in file but is in the ScanOptions.
     // Ignore it here, as projection will pad the batch with a null column.
```
This comment should go above L173 above not here

lidavidm commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -160,18 +160,30 @@ Status ResolveOneFieldRef(
     // Ignore it here, as projection will pad the batch with a null column.
     return Status::OK();
+  } else if (const FieldPath* path = field_ref.field_path()) {
+    int index = path->indices()[0];
```
Should we explicitly reject nested field paths here for now?

vibhatha replied:
That’s an unclear point to me yet. That’s why I also describe the current work as an  initial fix. How are we using nested fields so far? I was looking into it a bit.

lidavidm replied:
I mean, it's not fully implemented, so we should at least error or include a test case or both.

What do you mean by using? It's supported, but not very well, there are some prerequisite JIRAs needed to get things fully working but those have languished.

vibhatha replied:
I got your point. Let me error and add a few tests.

vibhatha commented:
@lidavidm I updated the test cases and added a check for the nested field paths. To add an error checking test, I think some of the util functions need to be changed to grab a returning `NotImplemented` status. The `SetFilter` function call triggers the error. Is it okay to return a `Status` from that method. I didn't do the change but would like to know if it is okay to do. I can proceed with it if it is okay. But is there a better alternative here?

westonpace commented on cpp/examples/arrow/dataset_parquet_scan_example.cc:
```diff
@@ -39,6 +39,17 @@ namespace ds = arrow::dataset;
 
 namespace cp = arrow::compute;
 
+/**
+ * @brief Run Example
+ * ./dataset-parquet-scan-example file:///sell_data.parquet
+ * sample data set
+ *           pickup_at dropoff_at  total_amount
+ *         0         A          B            10
+ *         1         B          A          1200
+ *         2         C          A          2400
+ *         3         A          C           500
+ */
+
```
Can you change the commenting here to be `//`?

westonpace commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -160,18 +160,35 @@ Status ResolveOneFieldRef(
+  };
+
+  if (const std::string* name = field_ref.name()) {
+    RETURN_NOT_OK(resolve_field_ref(name));
+    return Status::OK();
```
```suggestion
    return resolve_field_ref(name);
```

westonpace commented on cpp/examples/arrow/dataset_parquet_scan_example.cc:
```diff
@@ -62,8 +73,11 @@ struct Configuration {
   // make use of partition information and/or file metadata if possible.
-  cp::Expression filter =
-      cp::greater(cp::field_ref("total_amount"), cp::literal(1000.0f));
+  // Equivalent filter with field_name instead of field_index
+  // cp::Expression filter = cp::greater(cp::field_ref("total_amount"),
+  // cp::literal(1000.0f));
+  // 0: pickup_at, 1: dropoff_at, 2: total_amount
+  cp::Expression filter = cp::greater(cp::field_ref(2), cp::literal(1000.0f));
```
I think this comment makes the example a little more complex than needed.  Maybe we could do an AND expression `field_ref(2) > 1000 & field_ref("total_amount") < 2000` to demonstrate both types of field refs?  Or we could just leave it how it was and not worry about an example field ref by id.

vibhatha replied:
This is indeed a good one. Adding that.

westonpace commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -160,18 +160,35 @@ Status ResolveOneFieldRef(
+    int index = indices[0];
+    auto schema_field = manifest.schema_fields.at(index);
+    auto col_name = schema_field.field->name();
+    RETURN_NOT_OK(resolve_field_ref(&col_name));
+    return Status::OK();
```
```suggestion
    return resolve_field_ref(&col_name);
```

westonpace commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -160,18 +160,35 @@ Status ResolveOneFieldRef(
+                                    ", Nested FieldPaths are not supported!");
+    }
+    int index = indices[0];
+    auto schema_field = manifest.schema_fields.at(index);
```
Is `manifest.schema_fields` a vector?  If so can we use `[index]` instead of `.at(index)` for stylistic consistency?

westonpace reviewed (commented):
Looks good, I have just a few stylistic suggestions.

westonpace commented on cpp/examples/arrow/dataset_parquet_scan_example.cc:
```diff
@@ -39,6 +39,15 @@ namespace ds = arrow::dataset;
 
 namespace cp = arrow::compute;
 
+// @brief Run Example
```
```suggestion
// \brief Run Example
```

westonpace commented on cpp/examples/arrow/dataset_parquet_scan_example.cc:
```diff
@@ -62,9 +71,11 @@ struct Configuration {
-  cp::Expression filter =
-      cp::greater(cp::field_ref("total_amount"), cp::literal(1000.0f));
-
+  // Equivalent filter with field_name instead of field_index
```
```suggestion
// fields can be referenced by name or by index.  This example assumes the schema:
```

westonpace commented on cpp/examples/arrow/dataset_parquet_scan_example.cc:
```diff
@@ -62,9 +71,11 @@ struct Configuration {
 
   // Indicates the filter by which rows will be filtered. This optimization can
```
```suggestion
  // Indicates the expression by which rows will be filtered. This optimization can
```

westonpace commented on cpp/examples/arrow/dataset_parquet_scan_example.cc:
```diff
@@ -186,5 +197,9 @@ int main(int argc, char** argv) {
   auto table = GetTableFromScanner(scanner);
   std::cout << "Table size: " << table->num_rows() << "\n";
 
+  std::cout << "Data : " << std::endl;
```
```suggestion
  std::cout << "Data: " << std::endl;
```

westonpace commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -160,18 +160,34 @@ Status ResolveOneFieldRef(
+                                    ", Nested FieldPaths are not supported!");
+    }
+    int index = indices[0];
+    auto schema_field = manifest.schema_fields[index];
```
`manifest.schema_fields` will give us the fields in file order.  But, an integral field reference is actually a reference into the dataset schema.  So I think we want to be using the dataset schema here instead of the file manifest.

We should add a test case for this as it is a rather subtle point...

```
Parquet File 1:
x | y | z
1 | 2 | 3
```

```
Parquet File 2:
z | y | x
3 | 2 | 1
```

```
Dataset Schema:
 - x: int32
 - y: int32
 - z: int32
```

`field_ref(0)` should yield `[1, 1]`

vibhatha replied:
Thanks you for noting this one.

westonpace reviewed (changes requested):
Sorry for not realizing this earlier but I think we're indexing into the wrong schema.

lidavidm commented on cpp/src/arrow/dataset/file_parquet_test.cc:
```diff
@@ -390,6 +390,38 @@ TEST_P(TestParquetFileFormatScan, ScanRecordBatchReaderWithVirtualColumn) {
+TEST_P(TestParquetFileFormatScan, ScanRecordBatchReaderWithFieldPathFilter) {
+  TestScanWithFieldPathFilter();
+}
+TEST_P(TestParquetFileFormatScan, ScanRecordBatchReaderWithFieldPathMultiFileFilter) {
```
Can we make this a "common" test too?

vibhatha replied:
this is WIP test case @lidavidm I haven't finalized it yet. Still addressing the review.

lidavidm commented on cpp/src/arrow/dataset/test_util.h:
```diff
@@ -513,6 +513,33 @@ class FileFormatFixtureMixin : public ::testing::Test {
     EXPECT_OK_AND_ASSIGN(auto written, sink->Finish());
     return written;
   }
+  std::shared_ptr<Buffer> WriteToBufferFromTable(
```
nit: `WriteTableToBuffer`?

lidavidm commented on cpp/src/arrow/dataset/test_util.h:
```diff
@@ -513,6 +513,33 @@ class FileFormatFixtureMixin : public ::testing::Test {
     return written;
   }
+  std::shared_ptr<Buffer> WriteToBufferFromTable(
+      std::shared_ptr<Schema> schema,
```
A table already includes a schema

lidavidm commented on cpp/src/arrow/dataset/test_util.h:
```diff
@@ -823,6 +850,34 @@ class FileFormatScanMixin : public FileFormatFixtureMixin<FormatHelper>,
+  void TestScanWithFieldPathFilter() {
+    compute::Expression index_filter = equal(field_ref(0), literal(0));
+    compute::Expression name_filter = equal(field_ref("i64"), literal(0));
+    std::vector<int64_t> row_counts(2);
```
This is wrong when used with push_back, the end result will be a vector `[0, 0, x, y]`

lidavidm commented on cpp/src/arrow/dataset/test_util.h:
```diff
@@ -823,6 +850,34 @@ class FileFormatScanMixin : public FileFormatFixtureMixin<FormatHelper>,
+      for (auto maybe_batch : PhysicalBatches(fragment)) {
+        ASSERT_OK_AND_ASSIGN(auto batch, maybe_batch);
+        row_count += batch->num_rows();
+      }
```
IMO we should validate the schema of the batches too

lidavidm commented on cpp/src/arrow/dataset/file_parquet_test.cc:
```diff
@@ -390,6 +390,38 @@ TEST_P(TestParquetFileFormatScan, ScanRecordBatchReaderWithVirtualColumn) {
+
+  for (auto maybe_batch : PhysicalBatches(fragment)) {
+    ASSERT_OK_AND_ASSIGN(auto batch, maybe_batch);
+    std::cout << "Batch ::" << std::endl;
+    std::cout << batch->ToString() << std::endl;
```
We shouldn't print the data, also there doesn't seem to be an assertion in this test?

vibhatha replied:
This is part of addressing a review.

lidavidm commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -258,8 +275,8 @@ Result<std::vector<int>> InferColumnProjection(const parquet::arrow::FileReader&
   for (const auto& ref : field_refs) {
-    RETURN_NOT_OK(ResolveOneFieldRef(manifest, ref, field_lookup, duplicate_fields,
-                                     &columns_selection));
+    RETURN_NOT_OK(ResolveOneFieldRef(manifest, dataset_schema, ref, field_lookup,
```
nit, but I don't see a reason to create a copy of the dataset schema above, just pass it as `options.dataset_schema`

lidavidm commented on cpp/src/arrow/dataset/file_parquet.cc:
```diff
@@ -156,22 +156,37 @@ void AddColumnIndices(const SchemaField& schema_field,
 
 Status ResolveOneFieldRef(
-    const SchemaManifest& manifest, const FieldRef& field_ref,
+    const SchemaManifest& manifest, const std::shared_ptr<Schema> dataset_schema,
```
```suggestion
    const SchemaManifest& manifest, const std::shared_ptr<Schema>& dataset_schema,
```

amol- commented:
Closing because it has been untouched for a while, in case it's still relevant feel free to reopen and move it forward 👍
